### ISSUE TITLE
[FEA] Support `offset` parameter in `TakeOrderedAndProject`

### ISSUE THREAD

**Issue by sinkinben** [NVIDIA/spark-rapids#6034]
Parameter `offset` is introduced in `TakeOrderedAndProjectExec`, since spark-3.4.0.

```scala
case class TakeOrderedAndProjectExec(
    limit: Int,
    sortOrder: Seq[SortOrder],
    projectList: Seq[NamedExpression],
    child: SparkPlan,
    offset: Int = 0) extends UnaryExecNode { ... }
```

Correspondingly, we need to support it in spark-rapids.

```scala
    exec[TakeOrderedAndProjectExec](
      "Take the first limit elements as defined by the sortOrder, and do projection if needed",
      // The SortOrder TypeSig will govern what types can actually be used as sorting key data type.
      // The types below are allowed as inputs and outputs.
      ExecChecks((pluginSupportedOrderableSig + TypeSig.DECIMAL_128 +
          TypeSig.ARRAY + TypeSig.STRUCT + TypeSig.MAP).nested(), TypeSig.all),
      (takeExec, conf, p, r) =>
        new SparkPlanMeta[TakeOrderedAndProjectExec](takeExec, conf, p, r) {
          val sortOrder: Seq[BaseExprMeta[SortOrder]] =
            takeExec.sortOrder.map(GpuOverrides.wrapExpr(_, conf, Some(this)))
          val projectList: Seq[BaseExprMeta[NamedExpression]] =
            takeExec.projectList.map(GpuOverrides.wrapExpr(_, conf, Some(this)))
          override val childExprs: Seq[BaseExprMeta[_]] = sortOrder ++ projectList

          override def convertToGpu(): GpuExec = {
            // To avoid metrics confusion we split a single stage up into multiple parts but only
            // if there are multiple partitions to make it worth doing.
            val so = sortOrder.map(_.convertToGpu().asInstanceOf[SortOrder])
            if (takeExec.child.outputPartitioning.numPartitions == 1) {
              GpuTopN(takeExec.limit, so,
                projectList.map(_.convertToGpu().asInstanceOf[NamedExpression]),
                childPlans.head.convertIfNeeded())(takeExec.sortOrder)
            } else {
              GpuTopN(
                takeExec.limit,
                so,
                projectList.map(_.convertToGpu().asInstanceOf[NamedExpression]),
                GpuShuffleExchangeExec(
                  GpuSinglePartitioning,
                  GpuTopN(
                    takeExec.limit,
                    so,
                    takeExec.child.output,
                    childPlans.head.convertIfNeeded())(takeExec.sortOrder),
                  ENSURE_REQUIREMENTS
                )(SinglePartition)
              )(takeExec.sortOrder)
            }
          }
        }),
    
```

**Comment by revans2** [comment]
Just like with GlobalLimit we need to first make sure that we fall back to the CPU if offset is not 0.  That is a P1 for 3.4.0 support. After that we can look at implementing the functionality, but it is not as critical.

**Comment by revans2** [comment]
The current GpuTakeOrderedAndProject will keep the data in a single batch. This is because we made the assumption that the size of `limit` would be small.  But with an `offset` the number of rows we have to look at may be much larger. We have to keep around `offset + limit` because we cannot drop anything in the `offset` rows at the beginning of the sorted buffer until we know there is no way for it to bubble up to be a part of `limit` section. The only way we could know that is by knowing how many rows are left to be sorted.  This is because `offset` is often used for pagination. Look at the first 1000 rows. Then look at the next 1000 rows, and so on. 

This can probably be done as a follow on issue and should probably only happen if we see that the size of the data we keep ends up going over the target size for a batch.

**Comment by razajafri** [comment]
This is not a p0. It's a p1 I think and also per @revans2 comment

**Comment by razajafri** [comment]
@revans2 
I am confused about this issue. When I run the following in Spark-3.4.0 it seems to be on the GPU granted I don't see `TakeOrderedAndProjectExec` in the GPU plan but I do see `GpuTopN` and the entire plan is on the GPU after I set `spark.rapids.sql.exec.CollectLimitExec` to true and the results look fine. Is there work that's left to be done here?

```
scala> val df = Range(0,100).flatMap(i => Seq(i)).toDF() 
df: org.apache.spark.sql.DataFrame = [value: int]

scala> df.createOrReplaceTempView("test_table")

scala>  spark.conf.set("spark.rapids.sql.exec.CollectLimitExec", "true")

scala> spark.sql("select * from test_table order by value limit 10 offset 2").show()
23/04/28 17:14:55 WARN GpuOverrides: 
*Exec <CollectLimitExec> will run on GPU
  *Partitioning <SinglePartition$> will run on GPU
  *Exec <ProjectExec> will run on GPU
    *Expression <Alias> cast(value#1 as string) AS value#41 will run on GPU
      *Expression <Cast> cast(value#1 as string) will run on GPU
    *Exec <GlobalLimitExec> will run on GPU
      *Exec <ShuffleExchangeExec> will run on GPU
        *Partitioning <SinglePartition$> will run on GPU
        *Exec <SortExec> will run on GPU
          *Expression <SortOrder> value#1 ASC NULLS FIRST will run on GPU
          !Exec <ShuffleExchangeExec> cannot run on GPU because Columnar exchange without columnar children is inefficient
            @Partitioning <RangePartitioning> could run on GPU
              @Expression <SortOrder> value#1 ASC NULLS FIRST could run on GPU
                @Expression <AttributeReference> value#1 could run on GPU
            ! <LocalTableScanExec> cannot run on GPU because GPU does not currently support the operator class org.apache.spark.sql.execution.LocalTableScanExec
              @Expression <AttributeReference> value#1 could run on GPU

23/04/28 17:14:55 WARN GpuOverrides: 
*Exec <CollectLimitExec> will run on GPU
  *Partitioning <SinglePartition$> will run on GPU
  *Exec <ProjectExec> will run on GPU
    *Expression <Alias> cast(value#1 as string) AS value#41 will run on GPU
      *Expression <Cast> cast(value#1 as string) will run on GPU
    *Exec <GlobalLimitExec> will run on GPU
      *Exec <ShuffleExchangeExec> will run on GPU
        *Partitioning <SinglePartition$> will run on GPU
        *Exec <SortExec> will run on GPU
          *Expression <SortOrder> value#1 ASC NULLS FIRST will run on GPU
          !Exec <ShuffleExchangeExec> cannot run on GPU because Columnar exchange without columnar children is inefficient
            @Partitioning <RangePartitioning> could run on GPU
              @Expression <SortOrder> value#1 ASC NULLS FIRST could run on GPU
                @Expression <AttributeReference> value#1 could run on GPU
            ! <LocalTableScanExec> cannot run on GPU because GPU does not currently support the operator class org.apache.spark.sql.execution.LocalTableScanExec
              @Expression <AttributeReference> value#1 could run on GPU

23/04/28 17:14:55 WARN GpuOverrides: 
*Exec <CollectLimitExec> will run on GPU
  *Partitioning <SinglePartition$> will run on GPU
  *Exec <ProjectExec> will run on GPU
    *Expression <Alias> cast(value#1 as string) AS value#41 will run on GPU
      *Expression <Cast> cast(value#1 as string) will run on GPU
    *Exec <GlobalLimitExec> will run on GPU
      *Exec <ShuffleExchangeExec> will run on GPU
        *Partitioning <SinglePartition$> will run on GPU
        *Exec <SortExec> will run on GPU
          *Expression <SortOrder> value#1 ASC NULLS FIRST will run on GPU
          !Exec <ShuffleExchangeExec> cannot run on GPU because Columnar exchange without columnar children is inefficient
            @Partitioning <RangePartitioning> could run on GPU
              @Expression <SortOrder> value#1 ASC NULLS FIRST could run on GPU
                @Expression <AttributeReference> value#1 could run on GPU
            ! <LocalTableScanExec> cannot run on GPU because GPU does not currently support the operator class org.apache.spark.sql.execution.LocalTableScanExec
              @Expression <AttributeReference> value#1 could run on GPU

23/04/28 17:14:55 WARN GpuOverrides: 
!Exec <ShuffleExchangeExec> cannot run on GPU because Columnar exchange without columnar children is inefficient
  @Partitioning <RangePartitioning> could run on GPU
    @Expression <SortOrder> value#1 ASC NULLS FIRST could run on GPU
      @Expression <AttributeReference> value#1 could run on GPU
  ! <LocalTableScanExec> cannot run on GPU because GPU does not currently support the operator class org.apache.spark.sql.execution.LocalTableScanExec
    @Expression <AttributeReference> value#1 could run on GPU

23/04/28 17:14:55 WARN GpuOverrides: 
*Exec <CollectLimitExec> will run on GPU
  *Partitioning <SinglePartition$> will run on GPU
  *Exec <ProjectExec> will run on GPU
    *Expression <Alias> cast(value#1 as string) AS value#41 will run on GPU
      *Expression <Cast> cast(value#1 as string) will run on GPU
    *Exec <GlobalLimitExec> will run on GPU
      *Exec <ShuffleExchangeExec> will run on GPU
        *Partitioning <SinglePartition$> will run on GPU
        *Exec <SortExec> will run on GPU
          *Expression <SortOrder> value#1 ASC NULLS FIRST will run on GPU

23/04/28 17:14:55 WARN GpuOverrides: 
*Exec <ShuffleExchangeExec> will run on GPU
  *Partitioning <SinglePartition$> will run on GPU
  *Exec <SortExec> will run on GPU
    *Expression <SortOrder> value#1 ASC NULLS FIRST will run on GPU
    !Exec <AQEShuffleReadExec> cannot run on GPU because Unable to replace CustomShuffleReader due to child not being columnar

23/04/28 17:14:55 WARN GpuOverrides: 
*Exec <CollectLimitExec> will run on GPU
  *Partitioning <SinglePartition$> will run on GPU
  *Exec <ProjectExec> will run on GPU
    *Expression <Alias> cast(value#1 as string) AS value#41 will run on GPU
      *Expression <Cast> cast(value#1 as string) will run on GPU
    *Exec <GlobalLimitExec> will run on GPU

23/04/28 17:14:55 WARN GpuOverrides: 
*Exec <CollectLimitExec> will run on GPU
  *Partitioning <SinglePartition$> will run on GPU
  *Exec <ProjectExec> will run on GPU
    *Expression <Alias> cast(value#1 as string) AS value#41 will run on GPU
      *Expression <Cast> cast(value#1 as string) will run on GPU
    *Exec <GlobalLimitExec> will run on GPU

+-----+
|value|
+-----+
|    2|
|    3|
|    4|
|    5|
|    6|
|    7|
|    8|
|    9|
|   10|
|   11|
+-----+

```

**Comment by revans2** [comment]
@razajafri 

https://github.com/NVIDIA/spark-rapids/blob/54f77b084a6b23b68af5e980d0161cf8f4ef1f1a/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala#L3720-L3759

`TakeOrderedAndProject` internally does a shuffle to a single task as a part of the computation. When we designed the GPU replacement we split it up into TopN -> Shuffle -> TopN so that we didn't have to replicate a lot of the work we did for shuffle into our own TakeOrderedAndProject implementation. 

I don't know what is happening here because the plan does not appear to include a TakeOrderedAndProject with an offset in it. If Spark still supports offset in the code for TakeOrderedAndProject then we need to support it. We just need to spend some more time to understand in which conditions it will use the offset and in which cases it will not. If Spark no longer supports that code, then we don't need to implement it.
